engines/e_pkcs11.c

```diff
@@ -0,0 +1,771 @@
+    tmpl[0].type = CKA_CLASS;
+    tmpl[0].pValue = &key_class;
+    tmpl[0].ulValueLen = len_kc;
+    tmpl[1].type = CKA_KEY_TYPE;
+    tmpl[1].pValue = &key_type;
+    tmpl[1].ulValueLen = len_kt;
+    tmpl[2].type = CKA_ID;
+    tmpl[2].pValue = ctx->id;
+    tmpl[2].ulValueLen = id_len;
+
+    rv = pkcs11_funcs->C_FindObjectsInit(ctx->session, tmpl,
+                                         sizeof(tmpl) / sizeof(CK_ATTRIBUTE) );
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsInit failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_INIT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjects(ctx->session, &objhandle, 1, &count);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjects failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+    return objhandle;
+
+ err:
+    return 0;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    size_t vlen, i, j = 0;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    *value = OPENSSL_malloc(strlen(tmp) + 1);
+
+    if (*value == NULL) goto err;
+
+    end = strpbrk(tmp, ";");
+
+    BIO_snprintf(*value, end == NULL ? strlen(tmp) + 1 :
+             (size_t) (end - tmp + 1), "%s", tmp);
+    hex2bin = OPENSSL_malloc(strlen(*value) + 1);
+
+    if (hex2bin == NULL) goto err;
+
+    vlen = strlen(*value);
+    for (i = 0; i < vlen; i++) {
+        if (*(*value+i) == '%' && i < (strlen(*value)-2)) {
+            *(hex2bin+j) = pkcs11_hex_int(*(*value+i+1), *(*value+i+2));
+            i += 2;
+        } else {
+            *(hex2bin+j) = *(*value+i);
+        }
+        j++;
+    }
+    *(hex2bin+j) = '\0';
+    *value = hex2bin;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static int pkcs11_get_console_pin(char **pin)
+{
+    int i, ret = 0;
+
+#ifndef OPENSSL_NO_UI_CONSOLE
+
+    char *strbuf = NULL;
+
+    strbuf = OPENSSL_malloc(512);
+    for (;;) {
+        char prompt[200];
+        BIO_snprintf(prompt, sizeof(prompt), "Enter PIN: ");
+        strbuf[0] = '\0';
+        i = EVP_read_pw_string((char *)strbuf, 512, prompt, 1);
+        if (i == 0) {
+            if (strbuf[0] == '\0') {
+                goto err;
+            }
+            *pin = strbuf;
+            return 1;
+        }
+        if (i < 0) {
+            PKCS11_trace("bad password read\n");
+            goto err;
+        }
+    }
+
+ err:
+    OPENSSL_free(strbuf);
+#endif
+
+    return ret;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
```

Comment:
this function is too long, You better split-out  some of its parts in two or three sub-tasks ::
- arguments parsing, 
- session start/opening, 
 - private key loading ?

it will simplify the error management ode-path